File: ruby/lib/woothee/mobilephone.rb
# -*- coding: utf-8 -*-

require 'woothee/dataset'
require 'woothee/util'

module Woothee::MobilePhone
  extend Woothee::Util

  def self.challenge_docomo(ua, result)
    return false if ua.index('DoCoMo').nil? and ua.index(';FOMA;').nil?

    version = if ua =~ /DoCoMo\/[.0-9]+[ \/]([^- \/;()"']+)/o
                $1
              elsif ua =~ /\(([^;)]+);FOMA;/o
                $1
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('docomo'))
    update_version(result, version)
    true
  end

  def self.challenge_au(ua, result)
    return false if ua.index('KDDI-').nil?

    # fix Woothee::OS.challenge_mobile if you want to fix here
    version = if ua =~ /KDDI-([^- \/;()"']+)/o
                $1
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('au'))
    update_version(result, version)
    true
  end

  def self.challenge_softbank(ua, result)
    return false if ua.index('SoftBank').nil? and ua.index('Vodafone').nil? and ua.index('J-PHONE').nil?

    version = if ua =~ /(?:SoftBank|Vodafone|J-PHONE)\/[.0-9]+\/([^ \/;()]+)/o
                $1
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('SoftBank'))
    update_version(result, version)
    true
  end

  def self.challenge_willcom(ua, result)
    return false if ua.index('WILLCOM').nil? and ua.index('DDIPOCKET').nil?

    # fix Woothee::OS.challenge_mobile if you want to fix here
    version = if ua =~ /(?:WILLCOM|DDIPOCKET);[^\/]+\/([^ \/;()]+)/o
                $1
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('willcom'))
    update_version(result, version)
    true
  end

  def self.challenge_misc(ua, result)
    if ua.index('jig browser')
      update_map(result, Woothee::DataSet.get('jig'))
      if ua =~ /jig browser[^;]+; ([^);]+)/o
        update_version(result, $1)
      end
      return true
    end
    if ua.index('emobile/') or ua.index('OpenBrowser') or ua.index('Browser/Obigo-Browser')
      update_map(result, Woothee::DataSet.get('emobile'))
      return true
    end
    if ua.index('SymbianOS')
      update_map(result, Woothee::DataSet.get('SymbianOS'))
      return true
    end
    if ua.index('Hatena-Mobile-Gateway/')
      update_map(result, Woothee::DataSet.get('MobileTranscoder'))
      update_version(result, 'Hatena')
      return true
    end
    if ua.index('livedoor-Mobile-Gateway/')
      update_map(result, Woothee::DataSet.get('MobileTranscoder'))
      update_version(result, 'livedoor')
      return true
    end

    false
  end
end

File: python/lib/woothee/mobilephone.py
# -*- coding: utf-8 -*-

import re
from . import dataset
from . import util

def challenge_docomo(ua, result):
  if 'DoCoMo' not in ua and ';FOMA;' not in ua:
    return False
  obj = re.search('DoCoMo/[.0-9]+[ /]([^- /;()"\']+)', ua)
  if obj:
    version = obj.group(1)
  else:
    obj = re.search('\(([^;)]+);FOMA;', ua)
    if obj:
      version = obj.group(1)
    else:
      version = dataset.VALUE_UNKNOWN
  util.update_map(result, dataset.get('docomo'))
  util.update_version(result, version)
  return True

def challenge_au(ua, result):
  if 'KDDI-' not in ua:
    return False
  obj = re.search('KDDI-([^- /;()"\']+)', ua)
  if obj:
    version = obj.group(1)
  else:
    version = dataset.VALUE_UNKNOWN
  util.update_map(result, dataset.get('au'))
  util.update_version(result, version)
  return True

def challenge_softbank(ua, result):
  if 'SoftBank' not in ua and 'Vodafone' not in ua and 'J-PHONE' not in ua:
    return False
  obj = re.search('(?:SoftBank|Vodafone|J-PHONE)/[.0-9]+/([^ /;()]+)', ua)
  if obj:
    version = obj.group(1)
  else:
    version = dataset.VALUE_UNKNOWN
  util.update_map(result, dataset.get('SoftBank'))
  util.update_version(result, version)
  return True

def challenge_willcom(ua, result):
  if 'WILLCOM' not in ua and 'DDIPOCKET' not in ua:
    return False
  obj = re.search('(?:WILLCOM|DDIPOCKET);[^/]+/([^ /;()]+)', ua)
  if obj:
    version = obj.group(1)
  else:
    version = dataset.VALUE_UNKNOWN
  util.update_map(result, dataset.get('willcom'))
  util.update_version(result, version)
  return True

def challenge_misc(ua, result):
  if 'jig browser' in ua:
    util.update_map(result, dataset.get('jig'))
    obj = re.search('jig browser[^;]+; ([^);]+)', ua)
    if obj:
      util.update_version(result, obj.group(1))
    return True
  
  if 'emobile/' in ua or 'OpenBrowser' in ua or 'Browser/Obigo-Browser' in ua:
    util.update_map(result, dataset.get('emobile'))
    return True
  
  if 'SymbianOS' in ua:
    util.update_map(result, dataset.get('SymbianOS'))
    return True
  
  if 'Hatena-Mobile-Gateway/' in ua:
    util.update_map(result, dataset.get('MobileTranscoder'))
    util.update_version(result, 'Hatena')
    return True
  
  if 'livedoor-Mobile-Gateway/' in ua:
    util.update_map(result, dataset.get('MobileTranscoder'))
    util.update_version(result, 'livedoor')
    return True
  
  return False
